### ISSUE TITLE
[BUG] MATCH query not allowed on replicas

### ISSUE THREAD

**Issue by RobertConanMcMillan** [memgraph/memgraph#705]
**Memgraph version**
Experienced on 2.4.1 and 2.4.2

**Environment**
Tested on Docker with Neo4j.Driver (C#) and Memgraph Lab.

**Describe the bug**
The query
`MATCH (:Customer { id: 1 })-[:ATTACHED*1]->(group:Group) RETURN group;` 
returns
`Query failed: Write query forbidden on the replica!`
when run on a replica.

**To Reproduce**
1. No data is needed, query is determined to write before execution
2. Run the following query 'MATCH (:Customer { id: 1 })-[:ATTACHED*1]->(group:Group) RETURN group;'

**Expected behavior**
The query should be allowed to be executed on a replica, returning exactly the same as running the query on a main.

**Additional context**
The issue appears to be centered around relations, all queries with `-[:X]->` fail when run on a replica

**Comment by gitbuda** [comment]
Yep, this seems like a bug. I'll try to recreate and initiate the fix in the following days. @RobertConanMcMillan, what's the urgency on your side?

**Comment by RobertConanMcMillan** [comment]
Thanks for the quick reply!

We are starting a rolling release of memgraph into our production environment over the next couple of weeks, while this issue does not block that it's definitely less than optimal having all read-only clients using the main node.

A fix during the next couple of weeks would be much appreciated, but is manageable from our side until a fix is available

**Comment by gitbuda** [comment]
"We are starting a rolling release of memgraph into our production environment over the next couple of weeks..." -> very nice! Let me know if you need any support on that side 😄

This issue should be easy to fix because query type deduction is done during query parsing time with simple logic. Stay tuned for the fix!

**Comment by gitbuda** [comment]
@RobertConanMcMillan, the team jumped on this, will let you know as soon as we figure out what is going on CC @brunos252

**Comment by brunos252** [comment]
I have managed to recreate the issue. Its cause is the `query_execution->prepared_query->rw_type` is set to RW (which is wrong, and signals the replica that it can't do that) in `interpreter.cpp`. This assignment is done inside `read_write_type_checker.cpp`, code below.

Trying to simplify the query I found that this happens for:
`MATCH (c:Customer)-->() RETURN c;`
since its PROFILE is 
```
| "* Produce"     | 1               | "  7.437454 %"  | "  0.002012 ms" |
| "* Expand"      | 1               | "  4.647634 %"  | "  0.001257 ms" |
| "* Filter"      | 1               | " 30.805673 %"  | "  0.008333 ms" |
| "* ScanAll"     | 2               | " 56.793411 %"  | "  0.015363 ms" |
| "* Once"        | 2               | "  0.315829 %"  | "  0.000085 ms" |
```
For this query, the `rw_type` is RW.

And while running something like:
`PROFILE MATCH (c:Customer) RETURN c;` we get:
```
| "* Produce"     | 1               | "  7.479063 %"  | "  0.005671 ms" |
| "* Filter"      | 1               | " 32.547025 %"  | "  0.024677 ms" |
| "* ScanAll"     | 2               | " 59.649450 %"  | "  0.045226 ms" |
| "* Once"        | 2               | "  0.324462 %"  | "  0.000246 ms" |
```
The `rw_type` is R.

This type is updated in this function:
```
void ReadWriteTypeChecker::UpdateType(RWType op_type) {
  // Update type only if it's not the NONE type and the current operator's type
  // is different than the one that's currently inferred.
  if (type != RWType::NONE && type != op_type) {
    type = RWType::RW;
  }
  // Stop inference because RW is the most "dominant" type, i.e. it isn't
  // affected by the type of nodes in the plan appearing after the node for
  // which the type is set to RW.
  if (type == RWType::RW) {
    return;
  }
  if (type == RWType::NONE && op_type != RWType::NONE) {
    type = op_type;
  }
}
```

In the problematic query, this function is first called for `Expand` with `op_type` = `R`, while our `type` is still `NONE`. Then, it is called for Filter, with `op_type` = `NONE`, with our `type` being `R`, making it `RW` inside of the first `if` statement. 
Sometimes `Filter` is done before `Expand` and then there is no issue.

To generalize, When an operator with `rw_type=NONE` is called when the query `rw_type` is `R`, it is set to `RW`

The proposed fix is linked in the PR above

**Comment by gitbuda** [comment]
@RobertConanMcMillan, there is a plan to release v2.5 tomorrow, we probably won't be able to include this in the release (code review, testing, etc), but most likely, there will be a patch release in a week or so 😄

**Comment by gitbuda** [comment]
@RobertConanMcMillan, this Issue got automatically closed, the fix should have been done in #709, but it's not been released yet. I hope we can shepherd through [all the PRs marked with v2.5.1](https://github.com/memgraph/memgraph/pulls?q=is%3Apr+is%3Aopen+label%3Av2.5.1) next week. Stay tuned for it!
